src/lib/response-triggers/userlocation.ts

```diff
@@ -0,0 +1,334 @@
+import Trigger from '../../models/trigger';
+import { DbLocation, DbUsers } from '../database/db-interfaces';
+import Config from '../config-interface';
+import { MessageObject } from 'webex/env';
+import { Role, UserLocationRow } from '../../models/database';
+import { Command } from '../../models/command';
+import constants from '../../constants';
+import xmlMessageParser from '../parsers/xmlMessageParser';
+import { LocationAction } from '../../models/location-action';
+import { Argument } from '../../models/argument';
+import { Logger } from '../logger';
+import tableHelper from '../parsers/tableHelper';
+const stringWidth = require('string-width');
+
+export default class UserLocation extends Trigger {
+	dbUsers : DbUsers;
+	dbLocation : DbLocation;
+	config : Config;
+
+	constructor(dbUsers : DbUsers, dbLocation : DbLocation, config : Config) {
+		super();
+
+		this.dbUsers = dbUsers;
+		this.dbLocation = dbLocation;
+		this.config = config;
+	}
+
+	isToTriggerOn(message : MessageObject) : boolean {
+		let parsedMessage = xmlMessageParser.parseNonPegMessage(message);
+		return parsedMessage.botId === constants.botId && parsedMessage.command.toLowerCase().startsWith(Command.UserLocation);
+	}
+
+	async createMessage(message : MessageObject) : Promise<MessageObject> {
+		let args : Argument[];
+		try {
+			args = xmlMessageParser.parseOutArgs(message);
+		} catch(error) {
+			Logger.error(`[UserLocation.createMessage] Error parsing args: ${error.message}`);
+			return { markdown: `Error parsing request: ${error.message}` }
+		}
+
+		if (args.length < 3 || args[2].isMention) {
+			return { markdown: `Please specify a command. Possible values are ${Object.values(LocationAction).join(', ')}` };
+		}
+
+		let response : string;
+
+		if (this.config.checkRole(message.personId, Role.Admin) || this.config.checkRole(message.personId, Role.UserLocation)) {
+			response = await this.createMessageAdmin(args, message);
+		} else {
+			response = await this.createMessageNonAdmin(args, message);
+		}
+
+		return {
+			markdown: response
+		};
+	}
+
+	private async createMessageAdmin(args : Argument[], message : MessageObject) : Promise<string> {
+		switch(args[2].text.toLowerCase()) {
+			case LocationAction.Get:
+				return await this.getUserLocation(args, message.personId);
+			case LocationAction.Set:
+				return await this.setUserLocationAdmin(args, message);
+			case LocationAction.Delete:
+				return await this.deleteUserLocationAdmin(args, message);
+			default:
+				return `Unknown command. Possible values are ${Object.values(LocationAction).join(', ')}`;
+		}
+	}
+
+	private async createMessageNonAdmin(args : Argument[], message : MessageObject) : Promise<string> {
+		switch(args[2].text.toLowerCase()) {
+			case LocationAction.Get:
+				return await this.getUserLocation(args, message.personId);
+			case LocationAction.Set:
+				if (args.length !== 5) {
```

Comment:
Feels like these should be refactored out to methods.

Reply:
Done